docs/source/guides/data.rst

```diff
@@ -0,0 +1,562 @@
+
+        def setup(self, stage: Optional[str] = None):
+            # called on every process
+            vocab = load_vocab()
+            self.vocab_size = len(vocab)
+
+            self.train, self.val, self.test = load_datasets()
+            self.train_dims = self.train.next_batch.size()
+
+        def train_dataloader(self):
+            transforms = ...
+            return DataLoader(self.train, batch_size=64)
+
+        def val_dataloader(self):
+            transforms = ...
+            return DataLoader(self.val, batch_size=64)
+
+        def test_dataloader(self):
+            transforms = ...
+            return DataLoader(self.test, batch_size=64)
+
+        def predict_dataloader(self):
+            transforms = ...
+            return DataLoader(self.predict, batch_size=64)
+
+DataModules are easier to re-use compared to pure :class:`~torch.utils.data.Dataset` definitions.
+
+.. code-block:: python
+
+    # use an MNIST dataset
+    mnist_dm = MNISTDatamodule()
+    model = LitModel(num_classes=mnist_dm.num_classes)
+    trainer.fit(model, mnist_dm)
+
+    # or other datasets with the same model
+    imagenet_dm = ImagenetDatamodule()
+    model = LitModel(num_classes=imagenet_dm.num_classes)
+    trainer.fit(model, imagenet_dm)
+
+.. note:: :meth:`~pytorch_lightning.core.datamodule.LightningDataModule.prepare_data` is called on only one GPU in distributed training (automatically)
+.. note:: :meth:`~pytorch_lightning.core.datamodule.LightningDataModule.setup` is called on every GPU (automatically)
+
+Read :ref:`this <datamodules>` for more details on LightningDataModules.
+
+---------------
+
+*****************************************
+Dynamically Configuring Models using Data
+*****************************************
+
+When your models need to know about the data, it's best to process the data before passing it to the model.
+
+.. code-block:: python
+
+    # init dm AND call the processing manually
+    dm = ImagenetDataModule()
+    dm.prepare_data()
+    dm.setup()
+
+    model = LitModel(out_features=dm.num_classes, img_width=dm.img_width, img_height=dm.img_height)
+    trainer.fit(model, dm)
+
+
+1. use :meth:`~pytorch_lightning.core.datamodule.LightningDataModule.prepare_data` to download and process the :class:`~torch.utils.data.Dataset`.
+2. use :meth:`~pytorch_lightning.core.datamodule.LightningDataModule.setup` to do splits, and build your model internals
+
+An alternative to using a DataModule is to defer initialization of the models modules to the :meth:`~pytorch_lightning.core.datamodule.LightningDataModule.setup` method of your LightningModule as follows:
+
+.. testcode::
+
+    class LitMNIST(LightningModule):
+
+        def __init__(self):
+            self.l1 = None
+
+        def prepare_data(self):
+            download_data()
+            tokenize()
+
+        def setup(self, stage: Optional[str] = None):
+            # step is either 'fit', 'validate', 'test', or 'predict'. 90% of the time not relevant
```

Comment:
as a user, i dont know what "the 90% of the time not relevant" means. when is it relevant? and if its not relevant most of the time, why is it an argument at all? is https://github.com/PyTorchLightning/pytorch-lightning/issues/6420 a better approach?

use cases where stage today is relevant:
- If the stage is "fit", then I don't need to worry about setting up test data. this can be expanded for validate/test/predict, basically mapping trainer functions to running stages

are there others that we should highlight?